File: BatchTestTool/java/java automated tester/Test/compile_passed/solutions/EmailClient.java
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class EmailClient {
    private String addr;
    private int capacity;
    private List<Map<String, String>> inbox;

    public EmailClient(String addr, int capacity) {
        this.addr = addr;
        this.capacity = capacity;
        this.inbox = new ArrayList<>();
    }

    public boolean sendTo(EmailClient recv, String content, int size) {
        if (!recv.isFullWithOneMoreEmail(size)) {
            Map<String, String> email = new HashMap<>();
            email.put("sender", this.addr);
            email.put("receiver", recv.addr);
            email.put("content", content);
            email.put("size", String.valueOf(size));
            email.put("time", getCurrentTime());
            email.put("state", "unread");

            recv.inbox.add(email);
            return true;
        } else {
            this.clearInbox(size);
            return false;
        }
    }

    public Map<String, String> fetch() {
        if (inbox.isEmpty()) {
            return null;
        }

        for (Map<String, String> email : inbox) {
            if ("unread".equals(email.get("state"))) {
                email.put("state", "read");
                return email;
            }
        }

        return null;
    }

    public boolean isFullWithOneMoreEmail(int size) {
        int occupiedSize = getOccupiedSize();
        return occupiedSize + size > capacity;
    }

    public int getOccupiedSize() {
        int occupiedSize = 0;
        for (Map<String, String> email : inbox) {
            occupiedSize += Integer.parseInt(email.get("size"));
        }
        return occupiedSize;
    }

    public void clearInbox(int size) {
        if (addr.isEmpty()) return;

        int freedSpace = 0;
        while (freedSpace < size && !inbox.isEmpty()) {
            freedSpace += Integer.parseInt(inbox.get(0).get("size"));
            inbox.remove(0);
        }
    }

    public String getCurrentTime() {
        LocalDateTime now = LocalDateTime.now();
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-M-d H:m:s");
        return now.format(formatter);
    }

    public String getAddr() {
        return addr;
    }

    public void setAddr(String addr) {
        this.addr = addr;
    }

    public int getCapacity() {
        return capacity;
    }

    public void setCapacity(int capacity) {
        this.capacity = capacity;
    }

    public List<Map<String, String>> getInbox() {
        return inbox;
    }

    public void setInbox(List<Map<String, String>> inbox) {
        this.inbox = inbox;
    }
}

File: ClassEval_T/cpp/solution/EmailClient.cpp

#include <iostream>
#include <sstream>

#include <string>
#include <vector>
#include <map>
#include <ctime>

class EmailClient {
public:
    EmailClient(const std::string& addr, int capacity);
    bool send_to(EmailClient& recv, const std::string& content, int size);
    std::map<std::string, std::string> fetch();
    bool is_full_with_one_more_email(int size);
    int get_occupied_size();
    void clear_inbox(int size);
    std::string addr;
    int capacity;
    std::vector<std::map<std::string, std::string>> inbox;

    std::string get_current_time();
};


EmailClient::EmailClient(const std::string& addr, int capacity)
    : addr(addr), capacity(capacity) {}


std::string EmailClient::get_current_time() {
    time_t now = time(nullptr);
    tm* ltm = localtime(&now);
    std::ostringstream oss;
    oss << 1900 + ltm->tm_year << "-"
        << 1 + ltm->tm_mon << "-"
        << ltm->tm_mday << " "
        << ltm->tm_hour << ":"
        << ltm->tm_min << ":"
        << ltm->tm_sec;
    return oss.str();
}


bool EmailClient::send_to(EmailClient& recv, const std::string& content, int size) {
    if (!recv.is_full_with_one_more_email(size)) {
        std::map<std::string, std::string> email = {
            {"sender", this->addr},
            {"receiver", recv.addr},
            {"content", content},
            {"size", std::to_string(size)},
            {"time", get_current_time()},
            {"state", "unread"}
        };
        recv.inbox.push_back(email);
        return true;
    }
    else {
        this->clear_inbox(size);
        return false;
    }
}


std::map<std::string, std::string> EmailClient::fetch() {
    if (inbox.empty()) {
        return {};
    }
    for (auto& email : inbox) {
        if (email["state"] == "unread") {
            email["state"] = "read";
            return email;
        }
    }
    return {};
}


bool EmailClient::is_full_with_one_more_email(int size) {
    int occupied_size = get_occupied_size();
    return occupied_size + size > capacity;
}


int EmailClient::get_occupied_size() {
    int occupied_size = 0;
    for (const auto& email : inbox) {
        occupied_size += std::stoi(email.at("size"));
    }
    return occupied_size;
}


void EmailClient::clear_inbox(int size) {
    if (addr.empty()) return;

    int freed_space = 0;
    while (freed_space < size && !inbox.empty()) {
        freed_space += std::stoi(inbox.front()["size"]);
        inbox.erase(inbox.begin());
    }
}

